src/types/cart.ts

```diff
@@ -0,0 +1,14 @@
+import { Product } from "./product";
+
+export const ADD_TO_CART = 'ADD_TO_CART';
+
+export interface AddToCartAction {
+  type: typeof ADD_TO_CART,
+  product: Product,
+}
+
+export type CartActionTypes = AddToCartAction;
+
+export interface CartState {
```

Comment:
I think these kinds of types, where the state or action types are managed separately from the actions are one of the biggest pitfalls of Redux and Typescript, mainly because they introduce two layers of management - the expected types of actions / state shape, as well as the actions and reducers themselves. If you're not careful about typing things correctly, we've seen secondary definitions (like these) shift from the actual types, which can lead to losing type safety. As an example of when that can happen, if the reducer was not typed with `CartState` or an action was not typed with `CartActionTypes`, and instead relying on inference, it would be easy to miss an inconsistency. 

The good news is that all of these types can be inferred, though this does require helpers to accomplish that task. When you have them, all of these manual definitions can be removed, and it's a lot easier to work with Redux + typescript.